src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -193,115 +132,262 @@ object Skat {
+          if (w < 0)
+            fatal(s"Variant weights must be non-negative, got $w")
+          val x: Vector[Double] =
+            if (useDosages) {
+              RegressionUtils.dosages(gs, completeSamplesBc.value)
+            } else {
+              RegressionUtils.hardCalls(gs, n, sampleMaskBc.value)
+            }
+          keyIterator(key).map((_, (x, w)))
         case _ => Iterator.empty
       }
     }.groupByKey(), keyType)
   }
+  
+  def linearSkat(keyGsWeightRdd: RDD[(Annotation, Iterable[(Vector[Double], Double)])],
+    y: DenseVector[Double], cov: DenseMatrix[Double], maxSize: Int, accuracy: Double, iterations: Int): RDD[Row] = {
+    
+    val n = y.size
+    val k = cov.cols
+    val d = n - k
+
+    if (d < 1)
+      fatal(s"$n samples and $k ${ plural(k, "covariate") } including intercept implies $d degrees of freedom.")
+
+    // fit null model
+    val QR = qr.reduced.impl_reduced_DM_Double(cov)
+    val Qt = QR.q.t
+    val R = QR.r
+    val beta = R \ (Qt * y)
+    val res = y - cov * beta
+    val sigmaSq = (res dot res) / d
+
+    val sc = keyGsWeightRdd.sparkContext
+    val resBc = sc.broadcast(res)
+    val QtBc = sc.broadcast(Qt)
+    
+    def linearTuple(x: Vector[Double], w: Double): SkatTuple = {
+      val xw = x * math.sqrt(w)
+      val sqrt_q = resBc.value dot xw
+      SkatTuple(sqrt_q * sqrt_q, xw, QtBc.value * xw)
+    }
+    
+    val maxEntries = math.max(maxSize * maxSize, maxEntriesForSmallN)
+    
+    keyGsWeightRdd
+      .map { case (key, vs) =>
+        val vsArray = vs.toArray
+        val size = vsArray.length
+        if (size <= maxSize) {
+          val skatTuples = vsArray.map((linearTuple _).tupled)
+          val (q, gramian) = if (size.toLong * n <= maxEntries) {
+            computeGramianSmallN(skatTuples)
+          } else {
+            computeGramianLargeN(skatTuples)
+          }
+          // using q / sigmaSq since Z.t * Z = gramian / sigmaSq
+          val (pval, fault) = computePval(q / sigmaSq, gramian, accuracy, iterations)
+          
+          // returning qstat = q / (2 * sigmaSq) to agree with skat R package convention
+          Row(key, size, q / (2 * sigmaSq), pval, fault)
+        } else {
+          Row(key, size, null, null, null)
+        }
+      }
+  }
 
-  /*
-  Davies algorithm takes the eigenvalues of (G * sqrt(W)).t * P_0 * G * sqrt(W)
-  We evaluate this matrix as 0.5 * (A.t * A - B.t * B) where
-  linear:   A = G * sqrt(W)              B = Q.t * G * sqrt(W)
-  logistic: A = sqrt(V) * G * sqrt(W)    B = inv(L) * X.t * V * G * sqrt(W)
-  Here X = Q * R and L * L.t = X.t * V * X
-  */
-  def runSkatPerKey(st: Array[SkatTuple], skatStatScaling: Double): SkatStat = {
+  def logisticSkat(keyGsWeightRdd: RDD[(Any, Iterable[(Vector[Double], Double)])],
+    y: DenseVector[Double], cov: DenseMatrix[Double], maxSize: Int, accuracy: Double, iterations: Int): RDD[Row] = {
+
+    val n = y.size
+    val k = cov.cols
+    val d = n - k
```

Comment:
These three fields are defined twice in different methods. `accuracy` and `iterations` are passed into `linearSkat`, `logisticSkat`, and `computePval`. `maxSize` is passed to the first two.

This sounds like the situation where a class clarifies the logic by carrying common values used by multiple methods. I'm imagining something like

```scala
object Skat {
  def apply(...): KeyTable = { new Skat(...).keyTable() }
}

class KeyTable(vds: VariantDataset,
    variantKeys: String,
    singleKey: Boolean,
    weightExpr: String,
    yExpr: String,
    covExpr: Array[String],
    useDosages: Boolean,
    optMaxSize: Option[Int],
    accuracy: Double,
    iterations: Int) {
    private val maxSize = optMaxSize.getOrElse(Int.MaxValue)
    private val (y, cov, completeSampleIndex) = RegressionUtils.getPhenoCovCompleteSamples(vds, yExpr, covExpr)
    private val n = y.size
    private val k = cov.cols
    private val d = n - k

    if (d < 1)
      fatal(s"$n samples and $k ${ plural(k, "covariate") } including intercept implies $d degrees of freedom.")
    if (accuracy <= 0)
      fatal(s"tolerance must be positive, default is 1e-6, got $accuracy")    
    if (iterations <= 0)
      fatal(s"iterations must be positive, default is 10000, got $iterations")
    if (logistic) {
      val badVals = y.findAll(yi => yi != 0d && yi != 1d)
      if (badVals.nonEmpty)
        fatal(s"For logistic SKAT, phenotype must be Boolean or numeric with value 0 or 1 for each complete " +
          s"sample; found ${badVals.length} ${plural(badVals.length, "violation")} starting with ${badVals(0)}")
    }
    if (maxSize <= 0)
      fatal(s"Maximum group size must be positive, got $maxSize")

    def keyTable(): KeyTable = {
      val (keyGsWeightRdd, keyType) = toKeyGsWeightRdd()
      val skatSignature = TStruct(
        ("key", keyType),
        ("size", TInt32),
        ("qstat", TFloat64),
        ("pval", TFloat64),
        ("fault", TInt32))

      val skatRdd: RDD[Row] = 
        if (logistic)
          logisticSkat(keyGsWeightRdd)
        else
          linearSkat(keyGsWeightRdd)
    

      new KeyTable(vds.hc, skatRdd, skatSignature, Array("key"))
  }
}
```

I haven't fully thought through this, but I suspect if you restructure it along these lines you'll see some duplication disappear, and it nicely collects all the requirements in one place.